util/chunk/chunk.go

```diff
@@ -0,0 +1,382 @@
+	c.columns[colIdx].appendNameValue(set.Name, set.Value)
+}
+
+// AppendJSON appends a JSON value to the chunk.
+func (c *Chunk) AppendJSON(colIdx int, j json.JSON) {
+	c.columns[colIdx].appendInterface(j)
+}
+
+type column struct {
+	length     int
+	nullCount  int
+	nullBitmap []byte
+	offsets    []int32
+	data       []byte
+	elemBuf    []byte
+	ifaces     []interface{}
+}
+
+func (c *column) isFixed() bool {
+	return c.elemBuf != nil
+}
+
+func (c *column) isVarlen() bool {
+	return c.offsets != nil
+}
+
+func (c *column) isInterface() bool {
+	return c.ifaces != nil
+}
+
+func (c *column) reset() {
+	c.length = 0
+	c.nullCount = 0
+	c.nullBitmap = c.nullBitmap[:0]
+	c.offsets = c.offsets[:0]
+	c.data = c.data[:0]
+	c.ifaces = c.ifaces[:0]
+}
+
+func (c *column) isNull(rowIdx int) bool {
+	nullByte := c.nullBitmap[rowIdx/8]
+	return nullByte&(1<<(uint(rowIdx)&7)) == 0
+}
+
+func (c *column) setNullBitmap(on bool) {
+	idx := c.length >> 3
+	if idx >= len(c.nullBitmap) {
+		c.nullBitmap = append(c.nullBitmap, 0)
+	}
+	if on {
+		pos := uint(c.length) & 7
+		c.nullBitmap[idx] |= byte((1 << pos))
+	} else {
+		c.nullCount++
+	}
+}
+
+func (c *column) appendNull() {
+	c.setNullBitmap(false)
+	if c.isFixed() {
+		c.data = append(c.data, c.elemBuf...)
+	} else if c.isVarlen() {
+		c.offsets = append(c.offsets, c.offsets[c.length])
+	} else {
+		c.ifaces = append(c.ifaces, nil)
+	}
+	c.length++
+}
+
+func (c *column) finishAppendFixed() {
+	c.data = append(c.data, c.elemBuf...)
+	c.setNullBitmap(true)
+	c.length++
+}
+
+func (c *column) appendInt64(i int64) {
+	*(*int64)(unsafe.Pointer(&c.elemBuf[0])) = i
+	c.finishAppendFixed()
+}
+
+func (c *column) appendUint64(u uint64) {
+	*(*uint64)(unsafe.Pointer(&c.elemBuf[0])) = u
+	c.finishAppendFixed()
+}
+
+func (c *column) appendFloat32(f float32) {
+	*(*float32)(unsafe.Pointer(&c.elemBuf[0])) = f
+	c.finishAppendFixed()
+}
+
+func (c *column) appendFloat64(f float64) {
+	*(*float64)(unsafe.Pointer(&c.elemBuf[0])) = f
+	c.finishAppendFixed()
+}
+
+func (c *column) finishAppendVar() {
+	c.setNullBitmap(true)
+	c.offsets = append(c.offsets, int32(len(c.data)))
+	c.length++
+}
+
+func (c *column) appendString(str string) {
+	c.data = append(c.data, str...)
+	c.finishAppendVar()
+}
+
+func (c *column) appendBytes(b []byte) {
+	c.data = append(c.data, b...)
+	c.finishAppendVar()
+}
+
+func (c *column) appendInterface(o interface{}) {
+	c.ifaces = append(c.ifaces, o)
+	c.setNullBitmap(true)
+	c.length++
+}
+
+func (c *column) appendDuration(dur types.Duration) {
+	*(*types.Duration)(unsafe.Pointer(&c.elemBuf[0])) = dur
+	c.finishAppendFixed()
+}
+
+func (c *column) appendMyDecimal(dec *types.MyDecimal) {
+	*(*types.MyDecimal)(unsafe.Pointer(&c.elemBuf[0])) = *dec
+	c.finishAppendFixed()
+}
+
+func (c *column) appendNameValue(name string, val uint64) {
+	var buf [8]byte
+	*(*uint64)(unsafe.Pointer(&buf[0])) = val
+	c.data = append(c.data, buf[:]...)
+	c.data = append(c.data, name...)
+	c.finishAppendVar()
+}
```

Comment:
how about move `appendDuration` and `appendMyDecimal` together with the fixed length value append functions and move `appendNameValue` together with the variant length value append functions ?